ui/app/scripts/shared/services/parser.js

```diff
@@ -764,98 +801,133 @@ define(function() {
+						
+						// {"lines":[{"errors":null,"success":
+						// [{"group":"UNKNOWN_1","label":"time","type":"source","name":"time","options":{},"sourceChannelName":null,"sinkChannelName":null},{"group":"UNKNOWN_1","label":"log","type":"sink","name":"log","options":{},"sourceChannelName":null,"sinkChannelName":null}]
+						// }],"links":[]}
+						
+						var streamName = streamdef.name?streamdef.name:'UNKNOWN_'+lineNumber;
+						if (streamdef.apps) {
+							var alreadySeen = {};
+							for (var m=0;m<streamdef.apps.length;m++) {
+								var expectedType = 'processor';
+								if (m === 0 && !streamdef.sourceChannel) {
+									expectedType = 'source';
+								}
+								if (m === (streamdef.apps.length-1) && !streamdef.sourceChannel) {
+									expectedType = 'sink';
+								}
+								var sourceChannelName = null;
+								if (m === 0 && streamdef.sourceChannel) {
+									sourceChannelName = streamdef.sourceChannel.channel.name;
+								}
+								var sinkChannelName = null;
+								if (m===streamdef.apps.length-1 && streamdef.sinkChannel) {
+									sinkChannelName = streamdef.sinkChannel.channel.name;
+								}
+								app = streamdef.apps[m];
+								var uglyObject = {};
+								uglyObject.group=streamName;
+								if (app.label) {
+									uglyObject.label = app.label;
+								}
+								uglyObject.type = expectedType;
+								uglyObject.name = app.name;
+								uglyObject.range = {'start':{'ch':app.start,'line':lineNumber},'end':{'ch':app.end,'line':lineNumber}};
+								options = {};
+								optionsranges = {};
+								if (app.options) {
+									for (var o2=0;o2<app.options.length;o2++) {
+										option = app.options[o2];
+										options[option.name]=option.value;
+										optionsranges[option.name]={'start':{'ch':option.start,'line':lineNumber},'end':{'ch':option.end,'line':lineNumber}};
+									}
+								}
+								uglyObject.options=options;
+								uglyObject.optionsranges = optionsranges;
+								uglyObject.sourceChannelName = sourceChannelName;
+								uglyObject.sinkChannelName = sinkChannelName;
+								success.push(uglyObject);
+
+								var nameToCheck = uglyObject.label?uglyObject.label:uglyObject.name;
+								// Check that each app has a unique label (either explicit or implicit)
+								var previous = alreadySeen[nameToCheck];
+								if (typeof previous === 'number') {
+									recordError(streamdef, {
+										'msg': app.label ?
+											'Label \'' + app.label + '\' should be unique but app \'' + app.name + '\' (at position ' + m + ') and app \'' + streamdef.apps[previous].name + '\' (at position ' + previous + ') both use it'
+											: 'App \'' + app.name + '\' should be unique within the stream, use a label to differentiate multiple occurrences',
+										'range': uglyObject.range
+									});
+								} else {
+									alreadySeen[nameToCheck] = m;
+								}
+							}
+						} else {
+							// error case: ':stream:foo >'
```

Comment:
We want to communicate the information we do know back to the caller rather than give them nothing. Channels are attached to the apps they feed. So we need to create a fake app to hang the channel information off. That way we can supply it back to the caller and they will be able to show a node on the graph for the channel reference at least.  Errors will still be communicated back but at least they get to see the valid data we could parse.